core/SiteContentDetector.php

```diff
@@ -78,55 +88,82 @@ private function resetDetectionProperties() : void
+        // If site data was passed in, then just run the detection checks against it and return.
+        if ($siteData) {
+            $this->detectionChecks($detectContent);
+            return;
+        }
 
-            $url = Site::getMainUrlFor($idSite);
-            if (!$url) {
+        // Get the site id from the request object if not explicitly passed
+        if ($idSite === null) {
+            if (!isset($_REQUEST['idSite'])) {
                 return;
             }
-
-            try {
-                $siteData = Http::sendHttpRequestBy('curl', $url, $timeOut, null, null,
-                    null, 0, false, true);
-            } catch (\Exception $e) {
+            $idSite = Common::getRequestVar('idSite', null, 'int');
+            if (!$idSite) {
+                return;
             }
+        }
 
+        // Check and load previously cached site content detection data if it exists
+        $scdCacheId = 'SiteContentDetector_'.$idSite;
+
+        $siteContentDetectionIsCached = $this->cache->fetch($scdCacheId);
+        if ($siteContentDetectionIsCached) {
+            $this->consentManagerId = $this->cache->fetch($scdCacheId.'_consentManagerId');
+            $this->consentManagerName = $this->cache->fetch($scdCacheId.'_consentManagerName');
+            $this->consentManagerUrl = $this->cache->fetch($scdCacheId.'_consentManagerUrl');
+            $this->isConnected = $this->cache->fetch($scdCacheId.'_isConnected');
+            $this->ga3 = $this->cache->fetch($scdCacheId.'_ga3');
+            $this->ga4 = $this->cache->fetch($scdCacheId.'_ga4');
+            $this->gtm = $this->cache->fetch($scdCacheId.'_gtm');
+            return;
         }
 
+        // No cache hit, no passed data, so make a request for the site content
+        $siteData = $this->requestSiteData($idSite, $timeOut);
+
         // Abort if still no site data
         if ($siteData === null || $siteData === '') {
             return;
         }
 
         $this->siteData = $siteData;
-        $this->siteId = $idSite;
+
+        // We now have site data to analyze, so run the detection checks
+        $this->detectionChecks($detectContent);
+
+        // A request was made to get this data and it isn't currently cached, so write it to the cache now
+        $cacheLife = (60 * 60 * 24 * 7);
+        $this->cache->save($scdCacheId, 1, $cacheLife);
+        $this->cache->save($scdCacheId.'_consentManagerId', $this->consentManagerId, $cacheLife);
+        $this->cache->save($scdCacheId.'_consentManagerName', $this->consentManagerName, $cacheLife);
+        $this->cache->save($scdCacheId.'_consentManagerUrl', $this->consentManagerUrl, $cacheLife);
+        $this->cache->save($scdCacheId.'_isConnected', $this->isConnected, $cacheLife);
+        $this->cache->save($scdCacheId.'_ga3', $this->ga3, $cacheLife);
+        $this->cache->save($scdCacheId.'_ga4', $this->ga4, $cacheLife);
+        $this->cache->save($scdCacheId.'_gtm', $this->gtm, $cacheLife);
```

Comment:
Maybe we don't have usages like this yet but if `$detectContent` is not `ALL_CONTENT` then some of these caches would be incorrect. This would also impact the fetching logic. Perhaps it's better to cache them all as a single entry and then extract which ones we have, and if we don't have the `$detectContent` item we're after we need to request?

Not sure if relevant in real life (yet), but this tripped me up when testing with the example supplied in the `usage` doc string

Reply:
Good point :+1: I'll rework the caching logic to only use the cache if it has the required properties and only save the properties to the cache if a detection check was done.

Reply:
Nice I think that looks good. Incidentally we also have `Date::NUM_SECONDS_IN_DAY` and could add `Date::NUM_SECONDS_IN_WEEK` if needed for readability.